workflow/controller/operator.go

```diff
@@ -1001,6 +989,18 @@ func (woc *wfOperationCtx) processNodeRetries(node *wfv1.NodeStatus, retryStrate
 		return woc.markNodePhase(node.Name, lastChildNode.Phase, message), true, nil
 	}
 
+	if retryStrategy.Expression != "" && len(childNodeIds) > 0 {
```

Comment:
I was expecting this to move even further down the code or be refactored in some other way, so that we checked `retryStrategy` first.

For example - in the event you have a retryStrategy of `OnFailure`, but the node is in error, we'll still evaluate the expression.

What do you think?

Reply:
So instead of retry on [`Supplied Policy AND Expression`](https://argo-workflows.readthedocs.io/en/latest/retries/#retry-policies),
We want to refactor this into `Supplied Policy OR Expression`, is this right? I think this is the way to go

Reply:
No, I don't want to change it to `OR`. This is would be a non-backwards compatible change, and require version 4 of workflows.

Your original problem is that the expression is evaluated even if the policy wouldn't have allowed a retry. This is still the case, as far as I can tell, but in a more limited scenario.

I think the expression should only be evaluated if the policy would allow a retry.

Reply:
Make sense, updated